File: packages/ts/cairn-p2p/src/pairing/pin.ts
import { CairnError } from '../errors.js';
import { hkdfSha256 } from '../crypto/hkdf.js';

/** Crockford Base32 alphabet (excludes I, L, O, U). */
const CROCKFORD_ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

/** Pin code length in characters (before formatting). */
const PIN_LENGTH = 8;

/** HKDF info string for deriving the rendezvous ID from a pin code. */
const HKDF_INFO_PIN_RENDEZVOUS = new TextEncoder().encode('cairn-pin-rendezvous-v1');

/**
 * Generate a random 8-character Crockford Base32 pin code (40 bits entropy).
 * Returns the raw 8-character string (without formatting).
 */
export function generatePin(): string {
  const bytes = crypto.getRandomValues(new Uint8Array(5)); // 40 bits
  return encodeCrockford(bytes);
}

/**
 * Format a pin code as `XXXX-XXXX`.
 */
export function formatPin(pin: string): string {
  if (pin.length === PIN_LENGTH) {
    return `${pin.slice(0, 4)}-${pin.slice(4)}`;
  }
  return pin;
}

/**
 * Normalize a pin code input: uppercase, strip separators, apply Crockford substitutions.
 *
 * - Case-insensitive (uppercased)
 * - `I`/`L` -> `1`
 * - `O` -> `0`
 * - `U` removed (Crockford excludes U)
 * - Hyphens and spaces stripped
 */
export function normalizePin(input: string): string {
  return input
    .split('')
    .filter(c => c !== '-' && c !== ' ')
    .map(c => c.toUpperCase())
    .filter(c => c !== 'U')
    .map(c => {
      if (c === 'I' || c === 'L') return '1';
      if (c === 'O') return '0';
      return c;
    })
    .join('');
}

/**
 * Validate a normalized pin code: must be 8 Crockford Base32 characters.
 */
export function validatePin(normalized: string): void {
  if (normalized.length !== PIN_LENGTH) {
    throw new CairnError('PAIRING', `expected ${PIN_LENGTH} characters, got ${normalized.length}`);
  }
  for (const ch of normalized) {
    if (!CROCKFORD_ALPHABET.includes(ch)) {
      throw new CairnError('PAIRING', `invalid Crockford character: '${ch}'`);
    }
  }
}

/**
 * Derive a 32-byte rendezvous ID from a pin code.
 * Uses HKDF-SHA256 with info="cairn-pin-rendezvous-v1".
 */
export function derivePinRendezvousId(pinBytes: Uint8Array): Uint8Array {
  return hkdfSha256(pinBytes, new Uint8Array(0), HKDF_INFO_PIN_RENDEZVOUS, 32);
}

/** Encode 5 bytes (40 bits) to 8 Crockford Base32 characters. */
function encodeCrockford(bytes: Uint8Array): string {
  // Convert 5 bytes to a 40-bit integer (using bigint for precision)
  let bits = 0n;
  for (const b of bytes) {
    bits = (bits << 8n) | BigInt(b);
  }

  // Extract 8 x 5-bit chunks from the top
  let result = '';
  for (let i = 7; i >= 0; i--) {
    const index = Number((bits >> BigInt(i * 5)) & 0x1Fn);
    result += CROCKFORD_ALPHABET[index];
  }
  return result;
}

/** Decode a Crockford Base32 string (8 chars, normalized) to 5 bytes. */
export function decodeCrockford(input: string): Uint8Array {
  if (input.length !== PIN_LENGTH) {
    throw new CairnError('PAIRING', `expected ${PIN_LENGTH} characters, got ${input.length}`);
  }

  let bits = 0n;
  for (const ch of input) {
    const idx = CROCKFORD_ALPHABET.indexOf(ch);
    if (idx === -1) {
      throw new CairnError('PAIRING', `invalid Crockford character: '${ch}'`);
    }
    bits = (bits << 5n) | BigInt(idx);
  }

  // Extract 5 bytes from the 40-bit value
  const result = new Uint8Array(5);
  for (let i = 0; i < 5; i++) {
    result[i] = Number((bits >> BigInt((4 - i) * 8)) & 0xFFn);
  }
  return result;
}

File: packages/go/cairn-p2p/pairing/pin.go
package pairing

import (
	"crypto/rand"
	"fmt"
	"strings"
)

// Crockford Base32 alphabet (excludes I, L, O, U).
const crockfordAlphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

const pinLength = 8

// GeneratePin generates a random 8-character Crockford Base32 PIN code.
// Returns the formatted "XXXX-XXXX" string and the raw 5 bytes of entropy.
func GeneratePin() (string, []byte, error) {
	var bytes [5]byte // 40 bits of entropy
	if _, err := rand.Read(bytes[:]); err != nil {
		return "", nil, fmt.Errorf("pin entropy generation failed: %w", err)
	}
	raw := encodeCrockford(bytes)
	formatted := FormatPin(raw)
	return formatted, bytes[:], nil
}

// FormatPin formats an 8-character PIN as "XXXX-XXXX".
func FormatPin(pin string) string {
	if len(pin) == pinLength {
		return pin[:4] + "-" + pin[4:]
	}
	return pin
}

// NormalizePin normalizes a user-entered PIN:
//   - Uppercases
//   - Strips hyphens and spaces
//   - I/L -> 1, O -> 0
//   - Removes U (excluded from Crockford Base32)
func NormalizePin(input string) string {
	var b strings.Builder
	for _, c := range strings.ToUpper(input) {
		switch c {
		case '-', ' ':
			continue
		case 'U':
			continue
		case 'I', 'L':
			b.WriteByte('1')
		case 'O':
			b.WriteByte('0')
		default:
			b.WriteRune(c)
		}
	}
	return b.String()
}

// ValidatePin validates that a normalized PIN is a valid 8-character Crockford Base32 string.
func ValidatePin(normalized string) error {
	if len(normalized) != pinLength {
		return fmt.Errorf("invalid pin length: expected %d, got %d", pinLength, len(normalized))
	}
	for _, c := range normalized {
		if !strings.ContainsRune(crockfordAlphabet, c) {
			return fmt.Errorf("invalid pin character: '%c'", c)
		}
	}
	return nil
}

// encodeCrockford encodes 5 bytes (40 bits) to 8 Crockford Base32 characters.
func encodeCrockford(bytes [5]byte) string {
	var bits uint64
	for _, b := range bytes {
		bits = (bits << 8) | uint64(b)
	}
	result := make([]byte, 8)
	for i := 7; i >= 0; i-- {
		idx := (bits >> (uint(i) * 5)) & 0x1F
		result[7-i] = crockfordAlphabet[idx]
	}
	return string(result)
}

// DecodeCrockford decodes an 8-character Crockford Base32 string to 5 bytes.
func DecodeCrockford(input string) ([5]byte, error) {
	if len(input) != pinLength {
		return [5]byte{}, fmt.Errorf("expected %d characters, got %d", pinLength, len(input))
	}
	var bits uint64
	for _, ch := range input {
		idx := strings.IndexRune(crockfordAlphabet, ch)
		if idx < 0 {
			return [5]byte{}, fmt.Errorf("invalid crockford character: '%c'", ch)
		}
		bits = (bits << 5) | uint64(idx)
	}
	var result [5]byte
	for i := 4; i >= 0; i-- {
		result[4-i] = byte((bits >> (uint(i) * 8)) & 0xFF)
	}
	return result, nil
}
